src/syscall_mock/mod.rs

```diff
@@ -133,26 +136,45 @@ impl SyscallMock {
         Ok(path)
     }
 
-    pub fn handle_end(mut self, exitcode: i32) -> TestResult {
+    pub fn handle_end(mut self, exitcode: i32, captured_stderr: Option<Vec<u8>>) -> TestResult {
         if let Some(expected_step) = self.protocol.steps.pop_front() {
-            self.register_error(&expected_step.command.format(), "<script terminated>");
+            self.register_step_error(&expected_step.command.format(), "<script terminated>");
         }
         if exitcode != self.protocol.exitcode {
-            self.register_error(
+            self.register_step_error(
                 &format!("<exitcode {}>", self.protocol.exitcode),
                 &format!("<exitcode {}>", exitcode),
             );
         }
+        if let Some(expected_stderr) = &self.protocol.stderr {
+            match captured_stderr {
+                None => panic!("stderr expected, but not captured"),
+                Some(captured_stderr) => {
+                    if &captured_stderr != expected_stderr {
+                        self.register_error(format!(
+                            "  expected output to stderr: {:?}\
+                             \n  received output to stderr: {:?}\n",
+                            String::from_utf8_lossy(&expected_stderr.clone()).as_ref(),
+                            String::from_utf8_lossy(&captured_stderr.clone()).as_ref(),
+                        ));
+                    }
+                }
+            }
+        }
         self.result
     }
 
-    fn register_error(&mut self, expected: &str, received: &str) {
+    fn register_step_error(&mut self, expected: &str, received: &str) {
```

Comment:
I like the renaming here :slightly_smiling_face:

Style question: I agree that using borrowed values like `&str` is ideal when choosing a method's signature, but it seems like we are almost(?) always passing a `&format(..)` in as the parameter. 

Maybe this function could accept `String` values for `expected` and `received`? The benefit would be less syntax noise at the call sites, but that would mean this would be the terminal point in those values' lifetimes. Happy with either approach here, I'd say this is also a stylistic option :slightly_smiling_face:

Reply:
I tried that, it results in less `&`s but two more `.to_string()`s. Given that `&str` as an argument feels more natural, I'll go with that.